### PR TITLE
Purge queue refactor

### PR DESCRIPTION
**Pull request by taniabogatsch**
## DuckDB's Eviction Queue

DuckDB needs an eviction queue to track block handles to evicte when running close to the memory limit. It uses [this](https://github.com/cameron314/concurrentqueue) concurrent, lock-free implementation. The eviction strategy is LRU. We implement the strategy as described in the following paragraph.

Whenever a block handle has been used and can be evicted (no more readers), we add a `BufferEvictionNode` to the eviction queue (`AddToEvicationQueue`). To do so, we first increment the atomic timestamp of the block handle. Then, we construct a new `BufferEvictionNode` containing that timestamp and a `weak_ptr` to the block handle. That way, recently used block handles are in the front of the queue as we insert `BufferEvictionNodes` with their newer timestamps (LRU).

However, this approach also means that we accumulate 'dead' nodes in the queue, as we never remove nodes from the queue. Dead nodes are those with (1) destroyed block handles (we can no longer acquire the `shared_ptr` from the `weak_ptr`) or (2) nodes with older timestamps.

Thus, we create exactly one dead node if a block handle is destroyed (`~BlockHandle`) or when we insert a newer eviction node into the queue (`ts != 1`). This enables us to count the total dead nodes in the queue, even though we do not know where exactly they are.

To prevent the queue from growing indefinitely and to speed up finding a least recently used live node to evict, we need to purge these accumulated dead nodes from the queue regularly. Before this PR, our eviction queue increased slowly for long-running scenarios with a heavy query load, e.g., more than 150M queries. The purging algorithm could not keep up with some aggressive loads.

## Approach

- we bulk dequeue `purge_size` nodes in a single purge iteration. We count all nodes that we wrongly dequeue (alive nodes), and then bulk enqueue them back. We adjust the total dead nodes purged.

### When to trigger a purge iteration

There are two types of situations.
1. For most scenarios, purging `PURGE_SIZE_MULTIPLIER` more nodes than we insert is enough. This also counters oscillation for scenarios where most nodes are dead. If we always purge slightly more, we trigger a purge less often, as we purge below the trigger.
2. However, if the pressure on the queue becomes too contested, we need to purge more aggressively, i.e., we actively seek a specific number of dead nodes to purge. We count the total number of existing dead nodes as described above. We detect this situation by observing the queue's ratio between alive vs. dead nodes. If the ratio of alive vs. dead nodes grows faster than we can purge, we keep purging until we hit one of the following conditions.
2.1. We're back at an approximate queue size less than `purge_size * EARLY_OUT_MULTIPLIER`.
2.2. We're back at a ratio of `1*alive_node:(ALIVE_NODE_MULTIPLIER - 1)*dead_nodes`. We go below our initial
ratio of `1*alive_node:ALIVE_NODE_MULTIPLIER*dead_nodes` to decrease oscillation. 		
2.3. We've purged the entire queue: `max_purges` is zero. This is a worst-case scenario, guaranteeing that we always exit the loop.

### Other changes
- We no longer enter `PurgeQueue` during each `~BlockHandle`.
- Only one thread purges at any given time. This is governed by `purge_active`.

## Results

This is the output of our fast tests (`build/release/test/unittest`). We can see that we always hit the early-out condition after purging one iteration (approximately 8k elements).

```
-- PURGE --
approx queue size: 33206
approx dead nodes: 32180
RESTORED MINIMUM QUEUE SIZE
[1518/2920] (51%): test/sql/aggregate/distinct/grouped/multiple_grouping_sets...
-- PURGE --
approx queue size: 32768
approx dead nodes: 32735
RESTORED MINIMUM QUEUE SIZE
[1520/2920] (52%): test/sql/aggregate/distinct/grouped/combined_with_grouping...
-- PURGE --
approx queue size: 32768
approx dead nodes: 32725
RESTORED MINIMUM QUEUE SIZE
-- PURGE --
approx queue size: 32785
approx dead nodes: 32678
RESTORED MINIMUM QUEUE SIZE
[1979/2920] (67%): test/sql/storage/test_purge_queue.test                       
-- PURGE --
approx queue size: 32770
approx dead nodes: 32714
RESTORED MINIMUM QUEUE SIZE
-- PURGE --
approx queue size: 32777
approx dead nodes: 32710
RESTORED MINIMUM QUEUE SIZE
-- PURGE --
```

Here is another scenario where the ratio of alive vs. dead nodes is significantly higher. We only purge 8k elements once for the first few outputs, as we're still within the ratio (`RESTORED RATIO` is slightly misleading here). Then, when we initially hit that ratio, we must purge until we're below it again. Then, eventually, the queue stabilizes due to its LRU property.

```
-- PURGE --
approx queue size: 634483
approx dead nodes: 254721
RESTORED RATIO
-- PURGE --
approx queue size: 632607
approx dead nodes: 252555
RESTORED RATIO
[...]

-- PURGE --
approx queue size: 2400928
approx dead nodes: 1800918
RESTORED RATIO
-- PURGE --
approx queue size: 2402827
approx dead nodes: 1802821
-- PURGE --
approx queue size: 2402418
approx dead nodes: 1802412
-- PURGE --
approx queue size: 2401625
approx dead nodes: 1801621
-- PURGE --
approx queue size: 2400883
approx dead nodes: 1800880
-- PURGE --
approx queue size: 2400192
approx dead nodes: 1800189
RESTORED RATIO
```

cc @Mytherin @lnkuiper

### CODE REVIEW

File: src/storage/buffer/buffer_pool.cpp [hunk 1]
```diff
@@ -110,19 +111,63 @@ BufferPool::EvictionResult BufferPool::EvictBlocks(MemoryTag tag, idx_t extra_me
...
+
+	// defensive check, nothing to purge
+	if (approx_q_size < BULK_PURGE_SIZE) {
+		purge_active = false;
+		return;
+	}
+
+	vector<BufferEvictionNode> nodes;
+	nodes.resize(BULK_PURGE_SIZE);
+	idx_t alive_nodes = 0;
```
Comment (lnkuiper) [inline src/storage/buffer/buffer_pool.cpp:143]: Can pre-allocate in the header, could even be an array

**Review by lnkuiper — COMMENTED** [review]
Looks good, I have one small comment (also let's wait for regression tests)

File: src/include/duckdb/storage/buffer/buffer_pool.hpp [hunk 2]
```diff
@@ -86,12 +87,24 @@ class BufferPool {
...
 	unique_ptr<EvictionQueue> queue;
-	//! Total number of insertions into the eviction queue. This guides the schedule for calling PurgeQueue.
-	atomic<uint32_t> queue_insertions;
 	//! Memory manager for concurrently used temporary memory, e.g., for physical operators
 	unique_ptr<TemporaryMemoryManager> temporary_memory_manager;
 	//! Memory usage per tag
 	atomic<idx_t> memory_usage_per_tag[MEMORY_TAG_COUNT];
+
+	//! We trigger a purge of the eviction queue every INSERT_INTERVAL insertions
+	constexpr static idx_t INSERT_INTERVAL = DEFAULT_STANDARD_VECTOR_SIZE;
```
Comment (Mytherin) [inline src/include/duckdb/storage/buffer/buffer_pool.hpp:96]: `INSERT_INTERVAL` is now defined twice, once here and once in `BufferPool::AddToEvictionQueue` - with different values. Might be causing some confusion?

**Comment by taniabogatsch** [comment]
@Mytherin, I've updated the PR description to outline the current approach. This approach should be more resilient to different scenarios and never purge more than 8k nodes per purge, if not necessary, to reduce the ratio (avoid leaks). Maybe you can also have a look at the introduced constants.

File: src/include/duckdb/storage/buffer/buffer_pool.hpp [hunk 3]
```diff
@@ -86,12 +92,32 @@ class BufferPool {
...
+	//! exceed their allowed ratio. Must be greater than 1.
+	constexpr static idx_t ALIVE_NODE_MULTIPLIER = 4;
+
+	//! Total number of insertions into the eviction queue. This guides the schedule for calling PurgeQueue.
+	atomic<idx_t> evict_queue_insertions;
+	//! Total dead nodes in the eviction queue. There are two scenarios in which a node dies: (1) we destroy its block
+	//! handle, or (2) we insert a newer version into the eviction queue.
+	atomic<idx_t> total_dead_nodes;
+	//! Whether a queue purge is currently active. Only lets a single thread enter the purge phase.
+	atomic<bool> purge_active;
```
Comment (Mytherin) [inline src/include/duckdb/storage/buffer/buffer_pool.hpp:117]: Could we replace this with a regular mutex, using `try_lock`


https://en.cppreference.com/w/cpp/thread/mutex/try_lock#:~:text=Tries%20to%20lock%20the%20mutex,locked%20by%20any%20other%20thread.

File: src/storage/buffer/buffer_pool.cpp [hunk 4]
```diff
@@ -81,51 +92,166 @@ BufferPool::EvictionResult BufferPool::EvictBlocks(MemoryTag tag, idx_t extra_me
...
+	while (max_purges != 0) {
+
+		PurgeIteration(purge_size);
+
+		// update relevant sizes and potentially early-out
+		approx_q_size = queue->q.size_approx();
+
+		// early-out according to (2.1)
+		if (approx_q_size < purge_size * EARLY_OUT_MULTIPLIER) {
+			purge_active = false;
```
Comment (Mytherin) [inline src/storage/buffer/buffer_pool.cpp:237]: `break`

File: src/storage/buffer/buffer_pool.cpp [hunk 5]
```diff
@@ -81,51 +92,166 @@ BufferPool::EvictionResult BufferPool::EvictBlocks(MemoryTag tag, idx_t extra_me
...
+			return;
+		}
+
+		idx_t approx_dead_nodes = total_dead_nodes;
+		approx_dead_nodes = approx_dead_nodes > approx_q_size ? approx_q_size : approx_dead_nodes;
+		idx_t approx_alive_nodes = approx_q_size - approx_dead_nodes;
+
+		// early-out according to (2.2)
+		if (approx_alive_nodes * (ALIVE_NODE_MULTIPLIER - 1) > approx_dead_nodes) {
+			purge_active = false;
```
Comment (Mytherin) [inline src/storage/buffer/buffer_pool.cpp:247]: `break`

File: src/storage/buffer/buffer_pool.cpp [hunk 6]
```diff
@@ -81,51 +92,166 @@ BufferPool::EvictionResult BufferPool::EvictBlocks(MemoryTag tag, idx_t extra_me
...
+void BufferPool::PurgeQueue() {
+
+	// only one thread purges the queue, all other threads early-out
+	bool actual_purge_active;
+	do {
+		actual_purge_active = purge_active;
+		if (actual_purge_active) {
+			return;
+		}
+	} while (!std::atomic_compare_exchange_weak(&purge_active, &actual_purge_active, true));
```
[the remark this answers is not in the document]
Comment (Mytherin) [reply src/storage/buffer/buffer_pool.cpp:192]: Ideally we should use a lock_guard as well to ensure it always gets released when we exit the function, maybe there is some `try_lock_guard` or similar?

[the remark this answers is not in the document]
Comment (Mytherin) [reply src/storage/buffer/buffer_pool.cpp:192]: See here - https://en.cppreference.com/w/cpp/thread/lock_tag

You can use `try_lock` followed by pushing the mutex into a `lock_guard`

File: src/storage/buffer/buffer_pool.cpp [hunk 7]
```diff
@@ -81,51 +92,166 @@ BufferPool::EvictionResult BufferPool::EvictBlocks(MemoryTag tag, idx_t extra_me
...
+			purge_active = false;
+			return;
+		}
+
+		idx_t approx_dead_nodes = total_dead_nodes;
+		approx_dead_nodes = approx_dead_nodes > approx_q_size ? approx_q_size : approx_dead_nodes;
+		idx_t approx_alive_nodes = approx_q_size - approx_dead_nodes;
+
+		// early-out according to (2.2)
+		if (approx_alive_nodes * (ALIVE_NODE_MULTIPLIER - 1) > approx_dead_nodes) {
```
Comment (Mytherin) [inline src/storage/buffer/buffer_pool.cpp:246]: Do we still need -1 here?

**Review by Mytherin — COMMENTED** [review]
Thanks for all the fixes! Looks great - some more comments below, plus one comment above because reasons

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**